src/main/kotlin/com/atlassian/performance/tools/virtualusers/LoadTest.kt

```diff
@@ -111,85 +106,67 @@ internal class LoadTest(
             LinkedBlockingQueue<Runnable>(),
             ThreadFactoryBuilder().setNameFormat("virtual-user-%d").setDaemon(true).build()
         )
-        val stopSchedule = Executors.newScheduledThreadPool(
-            1,
-            ThreadFactoryBuilder().setNameFormat("deferred-stop").setDaemon(true).build()
-        )
-        val stop = stopSchedule.schedule(
-            {
-                val active = loadPool.activeCount
-                logger.info("Stopping load")
-                loadPool.shutdownNow()
-                ExploratoryVirtualUser.shutdown.set(true)
-                if (active != virtualUsers) {
-                    throw Exception("Expected $virtualUsers VUs to still be active, but encountered $active")
-                }
-            },
-            finish.toMillis(),
-            TimeUnit.MILLISECONDS
-        )
-        val deadline = now() + finish + maxStop
-        logger.info("Deadline for tests is $deadline.")
+        val segments = (1..virtualUsers).map { segmentLoad(it) }
+        segments.forEach { loadPool.submit { applyLoad(it) } }
+        Thread.sleep(finish.toMillis())
+        logger.info("Stopping load")
+        val active = loadPool.activeCount
+        segments.forEach { it.close() }
```

Comment:
Can we close it in parallel?

Reply:
Good idea, will do.

Reply:
OTOH, https://gist.github.com/dagguh/b663b34d74d15438b7c7d959c092c78e  :

```
2019-04-24T10:09:22,221Z ERROR {} [applying load #29dc17ce-c97b-4a5f-b7de-e320fc18001b] [ExploratoryVirtualUser] Done applying load
...
2019-04-24T10:13:52,627Z ERROR {} [applying load #423f06a9-45be-4afe-a9a7-8eae26153a0e] [ExploratoryVirtualUser] Done applying load
```
One of the `close`s took PT4M30S anyway, which is a lot of overhead for a PT20M load already. Parallel execution is still as slow as the slowest thread.

Nevertheless, the parallelisation is good, I only fear it's not enough.

Reply:
I changed the test to work with `.virtualUsers(n)`, where `n` can be greater than 1.
For `n = 12`, `overhead = PT1.033S`.
For `n = 120`, `overhead = PT1.548S`.
It's hard to find a good "greenifying" case for parallelization here. Unless we're fine with `driver.quit`, but that means we have a huge (uncontrolled) overhead anyway.

Reply:
https://github.com/atlassian/virtual-users/pull/9#discussion_r278467021
Parallelization is red-greened